jakarta-package-alignment/src/main/java/com/palantir/gradle/jakartapackagealignment/JakartaPackageAlignmentPlugin.java

```diff
@@ -15,24 +15,55 @@
  */
 package com.palantir.gradle.jakartapackagealignment;
 
+import com.google.common.collect.ImmutableSet;
+import java.util.Set;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.artifacts.component.ModuleComponentSelector;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
 
 public class JakartaPackageAlignmentPlugin implements Plugin<Project> {
     @Override
     public final void apply(Project project) {
-        project.getConfigurations().configureEach(configuration -> {
-            configuration.getResolutionStrategy().getDependencySubstitution().all(dep -> {
-                if (dep.getRequested() instanceof ModuleComponentSelector) {
-                    ModuleComponentSelector selector = (ModuleComponentSelector) dep.getRequested();
-                    VersionMappings.getReplacement(selector)
-                            .ifPresent(replacement -> dep.useTarget(
-                                    replacement,
-                                    "forced to Java EE 8 dependency because the requested Jakarta "
-                                            + "dependency is < Jakarta EE 9"));
-                }
+        project.getPluginManager().withPlugin("java-base", _plugin -> {
+            SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+            sourceSets.configureEach(sourceSet -> {
+                configureAllConfigurationsForSourceSet(project, sourceSet);
             });
         });
+
+        project.getPluginManager().withPlugin("com.palantir.consistent-versions", _plugin -> {
+            // necessary to get g-c-v to write substituted dependencies to the versions.lock file
+            project.getConfigurations().getByName("unifiedClasspath", this::configureConfigurationForSubstitution);
+        });
+    }
```

Comment:
Awareness of specific dependency plugins seems like a bug that may come back to haunt us. I don't know of a better option, but we should make sure we understand why this is necessary.

Reply:
Agreed, this feels hacky.

I just found this (see: https://docs.gradle.org/current/userguide/resolution_rules.html#sec:dependency_substitution_rules):
> Adding a dependency substitution rule to a configuration changes the timing of when that configuration is resolved. Instead of being resolved on first use, the configuration is instead resolved when the task graph is being constructed.

So I wonder if the dependency substitution rules on e.g. `runtimeClasspath` isn't getting picked up when `unifiedClasspath` is resolved because the dependencies/constraints are copied to that configuration before the task graph is constructed?